### PR TITLE
Don't bind expiration time to render callback

### PR DESCRIPTION
**Pull request by acdlite**
## Based on #16663 and #16678 

This is a fragile pattern because there's only meant to be a single task per root, running at a single expiration time. Instead of binding the expiration time to the render task, or closing over it, we should determine the correct expiration time to work on using fields we store on the root object itself.

This removes the "return a continuation" pattern from the `renderRoot` function. Continuation handling is now handled by the wrapper function, which I've renamed from `runRootCallback` to `performWorkOnRoot`. That function is merely an entry point to `renderRoot`, so I've also removed the callback argument.

So to sum up, at at the beginning of each task, `performWorkOnRoot` determines which expiration time to work on, then calls `renderRoot`. And before exiting, it checks if it needs to schedule another task.

### CODE REVIEW

File: packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.internal.js [hunk 1]
```diff
@@ -494,7 +494,7 @@ describe('ReactSuspenseWithNoopRenderer', () => {
         return <Text text="(empty)" />;
       }
       return (
-        <Suspense>
+        <Suspense fallback="Loading...">
```
Comment (acdlite) [inline packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.internal.js]: This test was testing the wrong thing, because a Suspense without a `fallback` will result in an error. So what this was actually testing was the "on error, try again at the lowest pending level."

File: packages/react-reconciler/src/__tests__/ReactIncrementalErrorHandling-test.internal.js [hunk 2]
```diff
@@ -256,56 +256,6 @@ describe('ReactIncrementalErrorHandling', () => {
     expect(ReactNoop.getChildren()).toEqual([span('Everything is fine.')]);
   });
 
-  it('on error, retries at a lower priority using the expiration of higher priority', () => {
```
Comment (acdlite) [inline packages/react-reconciler/src/__tests__/ReactIncrementalErrorHandling-test.internal.js]: Intentionally removed this behavior; I'll add another test to describe what the new semantics are.

File: packages/react-reconciler/src/__tests__/__snapshots__/ReactIncrementalPerf-test.internal.js.snap [hunk 3]
```diff
@@ -422,7 +422,10 @@ exports[`ReactDebugFiberPerf warns if an in-progress update is interrupted 1`] =
 `;
 
 exports[`ReactDebugFiberPerf warns if async work expires (starvation) 1`] = `
-"⚛ (Committing Changes)
+"⚛ (React Tree Reconciliation: Completed Root)
+  ⚛ Foo [mount]
```
Comment (acdlite) [inline packages/react-reconciler/src/__tests__/__snapshots__/ReactIncrementalPerf-test.internal.js.snap]: This was actually wrong before, because it improperly dropped the measurements from the render phase.

File: packages/react-reconciler/src/ReactFiberWorkLoop.js [hunk 4]
```diff
@@ -602,27 +585,43 @@ function ensureRootIsScheduled(root: FiberRoot) {
...
+// This is the entry point for every concurrent task.
+function performWorkOnRoot(root, isSync) {
+  // Determine the next expiration time to work on, using the fields stored
+  // on the root.
+  let expirationTime = getNextRootExpirationTimeToWorkOn(root);
+  if (expirationTime !== NoWork) {
+    if (expirationTime !== Sync) {
+      // Since we know we're in a React event, we can clear the current
+      // event time. The next update will compute a new event time.
+      currentEventTime = NoWork;
```
Comment (Jessidhia) [inline packages/react-reconciler/src/ReactFiberWorkLoop.js]: Is this a React event?

(is any queued async update considered an "event"?)

Comment (sebmarkbage) [reply packages/react-reconciler/src/ReactFiberWorkLoop.js]: The "event" here refers to a scheduled callback with the scheduler. I.e. clean stack.

File: packages/react-reconciler/src/ReactFiberWorkLoop.js [hunk 5]
```diff
@@ -1063,38 +1035,44 @@ function renderRoot(
...
-            // restarting now instead.
-            prepareFreshStack(root, expirationTime);
-            return renderRoot.bind(null, root, expirationTime);
+            const lastPingedTime = root.lastPingedTime;
+            if (lastPingedTime === NoWork || lastPingedTime >= expirationTime) {
+              // This render was pinged but we didn't get to restart earlier so
+              // try restarting now instead.
+              root.lastPingedTime = expirationTime;
+              prepareFreshStack(root, expirationTime);
+              return;
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberWorkLoop.js]: If `lastPingedTime === expirationTime` then I wouldn't expect there to be a callback already scheduled, right? So how does this pick up work back up?

File: packages/react-reconciler/src/ReactFiberWorkLoop.js [hunk 6]
```diff
@@ -1063,38 +1035,44 @@ function renderRoot(
...
 
-          const nextKnownPendingLevel = root.nextKnownPendingLevel;
-          if (nextKnownPendingLevel !== NoWork) {
-            // There's lower priority work. It might be unsuspended. Try rendering
-            // at that level.
-            return renderRoot.bind(null, root, nextKnownPendingLevel);
+          const nextTime = getNextRootExpirationTimeToWorkOn(root);
+          if (nextTime !== NoWork && nextTime !== expirationTime) {
+            // There's additional work on this root.
+            return;
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberWorkLoop.js]: It doesn't really hurt to schedule a callback that is just going to get canceled, right? Maybe it's ok to let it schedule a callback just in case we don't get called for a while to other work, we might as well commit?

File: packages/react-reconciler/src/ReactFiberWorkLoop.js [hunk 7]
```diff
@@ -1063,38 +1035,44 @@ function renderRoot(
...
             lastSuspendedTime !== NoWork &&
             lastSuspendedTime !== expirationTime
           ) {
-            // We should prefer to render the fallback of at the last
-            // suspended level.
-            return renderRoot.bind(null, root, lastSuspendedTime);
+            // We should prefer to render the fallback of at the last suspended
+            // level. Ping the last suspended level to try rendering it again.
+            root.lastPingedTime = lastSuspendedTime;
+            return;
```
Comment (sebmarkbage) [inline packages/react-reconciler/src/ReactFiberWorkLoop.js]: I'm paranoid about these not having a callback scheduled and us stalling.

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberWorkLoop.js]: I put `ensureRootIsScheduled` in the finally block of `performWorkOnRoot` to try to account for this.

I think it could theoretically happen for some of the early flush paths, which bypass `performWorkOnRoot`. But since they bypass `performWorkOnRoot`, they also won’t cause an existing callback to be canceled.

File: packages/react-reconciler/src/ReactFiberWorkLoop.js (hunk 6 above)
Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberWorkLoop.js]: https://github.com/facebook/react/pull/16715/files/6ace553825626774d9d9325be84389fe7fe5e33a..fd253c9b0662c48f9b780c6d48bc884c99cace7f#r322551325

File: packages/react-reconciler/src/ReactFiberWorkLoop.js (hunk 5 above)
Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberWorkLoop.js]: https://github.com/facebook/react/pull/16715/files/6ace553825626774d9d9325be84389fe7fe5e33a..fd253c9b0662c48f9b780c6d48bc884c99cace7f#r322551325

**Comment by acdlite** [comment]
Regarding your paranoia that we'll neglect to schedule a callback, I opened a follow-up PR that tries to prevent a refactor hazard: https://github.com/facebook/react/pull/16743

**Comment by acdlite** [comment]
Closed by https://github.com/facebook/react/commit/ab4951fc03750a726e412b90126e0737fbd04014
